executor/executor.go

```diff
@@ -122,35 +129,64 @@ func (e *Executor) Run(conflicts state.Keys, f func() error) {
 	e.tasks[id] = t
 
 	// Record dependencies
-	for k := range conflicts {
+	for k, v := range conflicts {
 		latest, ok := e.edges[k]
 		if ok {
-			lt := e.tasks[latest]
+			lt := e.tasks[latest.TaskID]
 			if !lt.executed {
 				if t.dependencies == nil {
 					t.dependencies = set.NewSet[int](defaultSetSize)
 				}
-				t.dependencies.Add(lt.id)
 				if lt.blocking == nil {
 					lt.blocking = set.NewSet[int](defaultSetSize)
 				}
-				lt.blocking.Add(id)
+				// key has ONLY a Read permission
+				if v == state.Read {
+					// lt contains either Allocate or Write permission. If lt had only a Read
+					// permission, no dependency or blocking would be added
+					if latest.Permissions.Has(state.Allocate) || latest.Permissions.Has(state.Write) {
+						t.dependencies.Add(lt.id) // t depends on lt to execute
+						lt.blocking.Add(id)       // lt is blocking this current task
+						// If lt is blocking any other txn, we need to record this
+						// to prevent subsequent txns that depend on lt from executing
+						// at the same time
+						e.blocking.Add(lt.id)
+					}
+				}
+
+				// key contains a Write permission
+				if v.Has(state.Write) {
+					// lt contains either Read, Allocate, or Write
+					if latest.Permissions.Has(state.Read) || latest.Permissions.Has(state.Allocate) || latest.Permissions.Has(state.Write) {
```

Comment:
I don't think it matters what permissions the dependencies have, we need to wait?

Or is the idea that we are "future-looking" and being explicit with our checks?